### PR TITLE
support new tsdbgw response with detailed info about invalid metrics

### PR DESCRIPTION
**Pull request by Dieterbe**
this should only be merged once https://github.com/raintank/tsdb-gw/pull/144 is merged and deployed to all our instances/clusters.

to test:
1) check out that tsdb-gw version, run `make` and `scripts/build_docker.sh` in the tsdbgw directory
2) you can use metrictank's docker stack, docker/docker-dev-custom-cfg-kafka/ but apply this change:
```
-    image: raintank/tsdb-gw:0.9.0
+    image: raintank/tsdb-gw:latest
```
3 ) using this branch, and after running `make`, run carbon-relay-ng with this config: https://gist.github.com/Dieterbe/2bdc5834a248b75df64377a70cb380d1

the important bits are disabling validation, reduced concurrency to have fewer requests with multiple points, and having the grafanaNet route for tsdbgw

you can then do this:
```
~ ❯❯❯ cat test.txt
my.series1;tag1=value1;tag2=value2 1 1
my.series2;tag1;value1;tag2=value2 1 1
my.series3;tag1=value1;tag2=value2 1 1
~ ❯❯❯ cat test.txt | nc localhost 2013
~ ❯❯❯ 
```
and will see:
```
2019-01-09 13:14:17.199 [WARNING] request contained 1 invalid metrics that were dropped (2 valid metrics were published in this request)
"invalid tag format" : 1 metrics.  Examples:
&schema.MetricData{Id:"1.6fb4b35dfce5647ed4758fcae062dc9a", OrgId:1, Name:"my.series2", Metric:"my.series2", Interval:10, Value:1, Unit:"unknown", Time:1, Mtype:"gauge", Tags:[]string{"tag1", "tag2=value2", "value1"}}
```

### CODE REVIEW

File: route/grafananet.go [hunk 1]
```diff
@@ -234,16 +236,37 @@ func (route *GrafanaNet) retryFlush(metrics []*schema.MetricData, buffer *bytes.
...
+		bod, err := ioutil.ReadAll(resp.Body)
 		resp.Body.Close()
+		if err != nil {
+			log.Warnf("GrafanaNet processed the request, but could not read the response: %s", err.Error())
+			return dur, nil
+		}
+		var resp MetricsResponse
+		err = json.Unmarshal(bod, &resp)
+		if err != nil {
+			log.Warnf("GrafanaNet processed the request, but could not parse the response: %s", err.Error())
```
Comment (DanCech) [inline route/grafananet.go]: Here and on line 250 the messages read as though the server couldn't read/parse the response.  Maybe something like `Received status code <resp.StatusCode> from server, but response body could not be <read/parsed>: <error>` would be clearer.

File: route/grafananet.go [hunk 2]
```diff
@@ -234,16 +236,37 @@ func (route *GrafanaNet) retryFlush(metrics []*schema.MetricData, buffer *bytes.
...
+			return dur, nil
+		}
+		if resp.Invalid != 0 {
+			var b strings.Builder
+			fmt.Fprintf(&b, "request contained %d invalid metrics that were dropped (%d valid metrics were published in this request)\n", resp.Invalid, resp.Published)
+			for key, vErr := range resp.ValidationErrors {
+				fmt.Fprintf(&b, "%q : %d metrics.  Examples:\n", key, vErr.Count)
```
Comment (DanCech) [inline route/grafananet.go]:
```suggestion
				fmt.Fprintf(&b, "  %q : %d metrics.  Examples:\n", key, vErr.Count)
```

File: route/grafananet.go [hunk 3]
```diff
@@ -234,16 +236,37 @@ func (route *GrafanaNet) retryFlush(metrics []*schema.MetricData, buffer *bytes.
...
+				for _, idx := range vErr.ExampleIds {
+					fmt.Fprintf(&b, "%#v\n", mda[idx])
```
Comment (DanCech) [inline route/grafananet.go]:
```suggestion
					fmt.Fprintf(&b, "   - %#v\n", mda[idx])
```

**Comment by Dieterbe** [comment]
a few more loose ends remaining internally before we can proceed with this.

**Pull request merged**